google/cloud/bigquery/client.py

```diff
@@ -2064,15 +2077,19 @@ def get_job(
             timeout=timeout,
         )
 
-        return self.job_from_resource(resource)
+        job_instance = self.job_from_resource(resource)  # never an UnknownJob
```

Comment:
Hm... I was under impression that `job_from_resource()` is a helper that only returns `UnknownJob` if somebody manually provides an "invalid"/incomplete resource dict.

However, if the backend can also return such resource (e.g. due to permissions), the assumption in the comment does not hold (although the return type in the docstring already excludes `UnknownJob`).

Do you maybe remember the combination of permissions that result in an `UnknownJob`?
(also, I might need assistance here, because my access to cloud console is now severely limited - I'll have to check)

Reply:
> Do you maybe remember the combination of permissions that result in an UnknownJob?
(also, I might need assistance here, because my access to cloud console is now severely limited - I'll have to check)

It related to creating a job with one account and trying to list all jobs for the project with another account.

See: `jobs.list` versus `jobs.listAlll` here: https://cloud.google.com/bigquery/docs/access-control




Permission | Description
-- | --
bigquery.jobs.list | List all jobs and retrieve metadata on any job submitted by any user.1 For jobs submitted by other users, details and metadata are redacted.
bigquery.jobs.listAll | List all jobs and retrieve metadata on any job submitted by any user.1

1 For any job you create, you automatically have the equivalent of the bigquery.jobs.get and bigquery.jobs.update permissions for that job.

Reply:
@tswast My access to Cloud Console is limited now and I cannot create an account anymore that only has `bigquery.jobs.list` permission (and possibly a few others to access jobs on the project), while the current development account that I use has the `bigquery.jobs.listAll` permission.

In other words, I don't think I can actually retrieve an `UnknownJob`.

Could you run the below script and see what happens if we call `get_job()` on an `UnknownJob`?

<details>
<summary>unknown_jobs.py</summary>

```py
from collections import Counter

from google.cloud import bigquery
from google.oauth2 import service_account


PROJECT_ID = "TODO: set"
DATASET_ID = "TODO: set"
TABLE_ID = "TODO: set"

MAIN_CRED_FILE = "TODO: /path/to/service_account.json"

# NOTE: This one must have "bigquery.jobs.list" permission, but NOT "bigquery.jobs.listAll".
# We want metadata of jobs submitted by other users redacted.
OTHER_CRED_FILE = "TODO: /path/to/other/service_account.json"


def main():
    credentials = service_account.Credentials.from_service_account_file(MAIN_CRED_FILE)
    client = bigquery.Client(credentials=credentials)

    # Create table if needed.
    table_name_full = f"{PROJECT_ID}.{DATASET_ID}.{TABLE_ID}"
    table = client.create_table(table_name_full, exists_ok=True)

    # Create at least one job (load job in this case).
    job_config = bigquery.LoadJobConfig(
        schema=[
            bigquery.SchemaField("user_id", "STRING"),
            bigquery.SchemaField("name", "STRING"),
        ],
        write_disposition=bigquery.WriteDisposition.WRITE_TRUNCATE,
    )
    row = {"user_id": "1234567890", "name": "John"}
    load_job = client.load_table_from_json([row], destination=table, job_config=job_config)
    result = load_job.result()
    print(result)

    print("=" * 20)

    # Using another client, use other users' jobs. The client should NOT have
    # "bigquery.jobs.listAll" permission.
    credentials = service_account.Credentials.from_service_account_file(OTHER_CRED_FILE)
    other_client = bigquery.Client(credentials=credentials)

    jobs_iter = other_client.list_jobs(
        project=PROJECT_ID, max_results=10_000, all_users=True
    )
    retrieved_jobs = list(jobs_iter)

    jobs_count = Counter(type(job) for job in retrieved_jobs)
    print(jobs_count)

    try:
        unknown_job = next(
            job for job in retrieved_jobs if isinstance(job, bigquery.UnknownJob)
        )
    except StopIteration:
        print("No UnknownJobs received, cannot test get_job().")
        return
    
    unknown_job = other_client.get_job(unknown_job)
    print(unknown_job._properties)


if __name__ == "__main__":
    main()
```
</details>

Reply:
FWIW, I preemptively changed the return type annotation to also include `UnknownJob`.